stl/inc/format

```diff
@@ -445,103 +449,239 @@ _NODISCARD constexpr bool _Is_execution_charset_utf8() {
 #pragma warning(pop)
 }
 
-inline constexpr bool _Is_execution_charset_utf8_v = _Is_execution_charset_utf8();
+inline constexpr char16_t _Width_estimate_low_intervals[] = { // Per N4885 [format.string.std]/11
+    0x1100u, 0x1160u, 0x2329u, 0x232Bu, 0x2E80u, 0x303Fu, 0x3040u, 0xA4D0u, 0xAC00u, 0xD7A4u, 0xF900u, 0xFB00u, 0xFE10u,
+    0xFE1Au, 0xFE30u, 0xFE70u, 0xFF00u, 0xFF61u, 0xFFE0u, 0xFFE7u};
 
-_NODISCARD constexpr int _Utf8_code_units_in_next_character(
-    const char* const _First, const char* const _Last) noexcept {
-    // Returns a count of the number of UTF-8 code units that compose the first encoded character in [_First, _Last),
-    // or -1 if [_First, _Last) doesn't contain an entire encoded character or *_First is not a valid lead byte.
-    const auto _Ch = static_cast<unsigned char>(*_First);
-    if (_Ch < 0b1000'0000u) {
-        return 1;
+inline constexpr char32_t _Width_estimate_high_intervals[] = { // Per N4885 [format.string.std]/11
+    0x1F300u, 0x1F650u, 0x1F900u, 0x1FA00u, 0x20000u, 0x2FFFEu, 0x30000u, 0x3FFFEu};
+
+template <auto& _Bounds>
+_NODISCARD constexpr int _Unicode_width_estimate(const char32_t _Ch) noexcept {
+    // Computes the width estimation for Unicode characters from N4885 [format.string.std]/11
+    int _Result = 1;
+    for (const auto& _Bound : _Bounds) {
+        if (_Ch < _Bound) {
+            return _Result;
+        }
+        _Result ^= 1;
     }
 
-    const auto _Len = static_cast<size_t>(_Last - _First);
+    return 1;
+}
+
+template <class _CharT, bool _Statically_Utf8 = _Is_execution_charset_utf8()>
+class _Fmt_codec;

```

Comment:
It is in the morning but I fail to see why we need this forward declaration?

Reply:
Or better said, why it is not together with the specializations

Reply:
This has to exist because it is the only declaration of the primary template. In my mind it _is_ together with the specializations, but the "private base should immediately precede derived" rule binds tighter since `_Fmt_codec_base` is a base of only the `_Fmt_codec<char, meow>` partial specialization.

I don't think there's a convention here - I'm just making it up - so feel free to ask me to move things and I will. This definitely is not a hill I plan to die on.

Reply:
The fact that you can articulate a reason while I am here with "this looks strange" is quite a good reason to keep it as it is.